src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -410,267 +584,458 @@ private string GenerateForObject(TypeMetadata typeMetadata)
+                }
+                else
+                {
+                    runtimePropName = clrPropName;
+                }
+
+                return runtimePropName;
+            }
+
+            private string GenerateForType(TypeGenerationSpec typeMetadata, string metadataInitSource, string? additionalSource = null)
+            {
+                string typeCompilableName = typeMetadata.TypeRef;
+                string typeFriendlyName = typeMetadata.TypeInfoPropertyName;
+                string typeInfoPropertyTypeRef = $"{JsonTypeInfoTypeRef}<{typeCompilableName}>";
+
+                return @$"{GetJsonContextDeclarationSource()}
 {{
-    {JsonContextDeclarationSource}
+    private {typeInfoPropertyTypeRef} _{typeFriendlyName};
+    public {typeInfoPropertyTypeRef} {typeFriendlyName}
     {{
-        private JsonTypeInfo<{typeCompilableName}> _{typeFriendlyName};
-        public JsonTypeInfo<{typeCompilableName}> {typeFriendlyName}
+        get
         {{
-            get
+            if (_{typeFriendlyName} == null)
             {{
-                if (_{typeFriendlyName} == null)
-                {{
-                    {WrapWithCheckForCustomConverterIfRequired(metadataInitSource, typeCompilableName, typeFriendlyName, GetNumberHandlingAsStr(typeMetadata.NumberHandling))}
-                }}
-
-                return _{typeFriendlyName};
+                {WrapWithCheckForCustomConverterIfRequired(metadataInitSource, typeCompilableName, typeFriendlyName, GetNumberHandlingAsStr(typeMetadata.NumberHandling))}
             }}
-        }}{additionalSource}
-    }}
-}}
-";
+
+            return _{typeFriendlyName};
+        }}
+    }}{additionalSource}
+}}";
             }
 
             private string WrapWithCheckForCustomConverterIfRequired(string source, string typeCompilableName, string typeFriendlyName, string numberHandlingNamedArg)
             {
-                if (!_honorRuntimeProvidedCustomConverters)
+                if (_currentContext.SerializerOptions.IgnoreRuntimeCustomConverters)
                 {
                     return source;
                 }
 
-                return @$"JsonConverter customConverter;
-                    if ({OptionsInstanceVariableName}.Converters.Count > 0 && (customConverter = {RuntimeCustomConverterFetchingMethodName}(typeof({typeCompilableName}))) != null)
-                    {{
-                        _{typeFriendlyName} = {JsonMetadataServicesClassName}.{GetCreateValueInfoMethodRef(typeCompilableName)}({OptionsInstanceVariableName}, customConverter);
-                    }}
-                    else
-                    {{
-                        {source.Replace(Environment.NewLine, $"{Environment.NewLine}    ")}
-                    }}";
+                return @$"{JsonConverterTypeRef} customConverter;
+                if ({OptionsInstanceVariableName}.Converters.Count > 0 && (customConverter = {RuntimeCustomConverterFetchingMethodName}(typeof({typeCompilableName}))) != null)
```

Comment:
(nit) the whitespace is off in this method. The generated code looks like:

```C#
                if (_DictionarySystemStringSystemBoolean == null)
                {
                    global::System.Text.Json.Serialization.JsonConverter customConverter;
                        if (Options.Converters.Count > 0 && (customConverter = GetRuntimeProvidedCustomConverter(typeof(global::System.Collections.Generic.Dictionary<System.String,System.Boolean>))) != null)
                        {
                            _DictionarySystemStringSystemBoolean = global::System.Text.Json.Serialization.Metadata.JsonMetadataServices.CreateValueInfo<global::System.Collections.Generic.Dictionary<System.String,System.Boolean>>(Options, customConverter);
                        }
                        else
                        {
                            _DictionarySystemStringSystemBoolean = global::System.Text.Json.Serialization.Metadata.JsonMetadataServices.CreateDictionaryInfo<global::System.Collections.Generic.Dictionary<System.String,System.Boolean>, global::System.String, global::System.Boolean>(Options, () => new global::System.Collections.Generic.Dictionary<global::System.String, global::System.Boolean>(), this.String, this.Boolean, default, serializeFunc: DictionarySystemStringSystemBooleanSerialize);
                        }
                }
        
                return _DictionarySystemStringSystemBoolean;
            }
        }
```

notice that the whitespace is off here:

![image](https://user-images.githubusercontent.com/8291187/119687077-84b0db00-be0c-11eb-957b-a4e055a2024c.png)